lib/brakeman/report/report_sarif.rb

```diff
@@ -0,0 +1,112 @@
+require 'brakeman/report/report_json'
+
+class Brakeman::Report::SARIF < Brakeman::Report::JSON
+  def generate_report
+    sarif_log = {
+      :version => '2.1.0',
+      :$schema => 'https://schemastore.azurewebsites.net/schemas/json/sarif-2.1.0-rtm.4.json',
+      :runs => runs,
+    }
+    JSON.pretty_generate sarif_log
+  end
+
+  def runs
+    [
+      {
+        :tool => {
+          :driver => {
+            :name => 'Brakeman',
+            :informationUri => 'https://brakemanscanner.org',
+            :rules => rules,
+          },
+        },
+        :artifacts => artifacts,
+        :results => results,
+      },
+    ]
+  end
+
+  def rules
+    @rules ||= unique_warnings.map do |warning|
+      check_name = warning.check.gsub(/^Brakeman::Check/, '')
+      check_description = check_descriptions[check_name]
+      {
+        :id => warning.warning_code.to_s,
+        :name => warning.warning_type,
+        :shortDescription => {
+          :text => check_description,
+        },
+        :fullDescription => {
+          :text => check_description,
+        },
+        :helpUri => warning.link,
+        :properties => {
+          :warningType => warning.warning_type,
+          :checkName => check_name,
+          :tags => [check_name],
+        },
+      }
+    end
+  end
+
+  def artifacts
+    @artifacts ||= unique_locations.map do |location|
+      {
+        :location => {
+          :uri => location,
+          :uriBaseId => '%SRCROOT%',
+        }
+      }
+    end
+  end
+
+  def results
+    @results ||= all_warnings.map do |warning|
+      rule_id = warning.warning_code.to_s
+      result = {
+        :level => 'warning',
+        :message => {
+          :text => warning.message.to_s,
+        },
+        :locations => [
+          :physicalLocation => {
+            :artifactLocation => {
+              :uri => warning.file.relative,
+              :uriBaseId => '%SRCROOT%',
+              :index => unique_locations.index { |l| l == warning.file.relative },
+            },
+          }
+        ],
+        :ruleId => rule_id,
+        :ruleIndex => rules.index { |r| r[:id] == rule_id },
+      }
+
+      # Include region in location where applicable
+      if warning.line.is_a? Integer
+        # TODO: we may be able to derive startColumn based on user_input
```

Comment:
I'm happy to take a 🔪 at this, but do you think this sounds reasonable, @presidentbeef? The plan is to take the `user_input` values and extract the start / end column locations in the code where they appear. The `user_input` should then be highlighted appropriately in the GitHub UI.

Reply:
Might be difficult. RubyParser doesn't have column information and Ruby2Ruby (used for output formatting) doesn't necessarily return exactly the same code as the input. There would be a lot of guesswork involved.

One of the issues is that what Brakeman "sees" is not the same as the original code.

Reply:
Sounds tricky! I think I'll put this idea on ice for now then!